### PR TITLE
Sparse/on-demand connections

### PR DESCRIPTION
**Pull request by edenhill**
* Sparse/On-demand broker connections (enable with `enable.sparse.connections=true` - currently disabled by default)
 * Formalizes, cleanups up, and optimizes broker handler thread Ops/IO waiting.
 * Removes all internal uses of socket.blocking.max.ms, this property is now a no-op
 * Bug fix: Producer did not round-robin partitions.

Big thanks to @gduranceau for his work on the [Connection on demand to brokers PR](https://github.com/edenhill/librdkafka/pull/1659).

TODO: The test suite passes with sparse connections enabled, but a sparse.conn specific test needs to be added.

### CODE REVIEW

File: INTRODUCTION.md [hunk 1]
```diff
@@ -794,6 +794,64 @@ error log is silenced (will only be seen with debug enabled).
 but it is recommended to instead rely on the above heuristics.
 
 
+#### Sparse connections
+
+Maintaining an open connection to each broker in the cluster is problematic
+for clusters with a large number of brokers, as well with clusters with a
```
Comment (mhowlett) [inline INTRODUCTION.md]: 'as well as for'

File: INTRODUCTION.md [hunk 2]
```diff
@@ -794,6 +794,64 @@ error log is silenced (will only be seen with debug enabled).
...
+large number of clients.
+
+To alleviate this the `enable.sparse.connections=true` configuration property
```
Comment (mhowlett) [inline INTRODUCTION.md]: maybe: "To alleviate this, the `enable.sparse.connections` configuration property can
be set to `true`, in which case the client will only connect to brokers it needs
to commnicate with, and only when necessary."

File: INTRODUCTION.md [hunk 3]
```diff
@@ -794,6 +794,64 @@ error log is silenced (will only be seen with debug enabled).
 but it is recommended to instead rely on the above heuristics.
 
 
+#### Sparse connections
```
Comment (mhowlett) [inline INTRODUCTION.md]: under Brokers section above, maybe the first line is better: "On initialization, librdkafka only needs a partial list of brokers (at least one), called the bootstrap brokers".

File: src/rdkafka_broker.h [hunk 4]
```diff
@@ -198,6 +205,30 @@ struct rd_kafka_broker_s { /* rd_kafka_broker_t */
...
+         *   on-demand mode when they all reach zero.
+         *   After incrementing any of the counters a broker wakeup
+         *   should be signalled to expedite handling. */
+        struct {
+                /**< Producer: partitions are being produced to.
+                 *   Consumer: partitions are being fetched from.
+                 *
+                 *   Counter is maintained by the broker handler thread
+                 *   itself, no need for atomic/locking.
+                 *   Maybe be reset to 0 on each producer|consumer_serve() loop
```
Comment (mhowlett) [inline src/rdkafka_broker.h]: 'Maybe be ...' has me completely confused.

File: src/rdkafka.c [hunk 5]
```diff
@@ -1511,6 +1520,13 @@ static void rd_kafka_stats_emit_all (rd_kafka_t *rk) {
 static void rd_kafka_topic_scan_tmr_cb (rd_kafka_timers_t *rkts, void *arg) {
         rd_kafka_t *rk = rkts->rkts_rk;
 	rd_kafka_topic_scan_all(rk, rd_clock());
+
+        /* Sparse connections:
+         * try to maintain at least one connection to the cluster. */
+        if (rk->rk_conf.sparse_connections &&
```
Comment (mhowlett) [inline src/rdkafka.c]: seems unrelated to the purpose of the function?
perhaps consider changing the name of the _cb or having more than one _cb

File: src/rdkafka_broker.h (hunk 4 above)
Comment (edenhill) [reply src/rdkafka_broker.h]: Mission accomplished

**Comment by edenhill** [comment]
I think we'll want to make sparse connections the default right away, to get as much pre-release coverage as possible, and leave the disable-mode as a workaround, to eventually remove it alltogether.
I don't see a case where the disable-mode is actually needed.

**Comment by edenhill** [comment]
@mhowlett @rnpridgeon (and whoever else wants to help out!) Please review the exponential backoff commit.

File: src/rdkafka_conf.c [hunk 6]
```diff
@@ -403,8 +403,23 @@ static const struct rd_kafka_property rd_kafka_properties[] = {
...
+          "**Deprecated**: No longer used. See `reconnect.backoff.ms` and "
+          "`reconnect.backoff.max.ms`.",
+          0, 60*60*1000, 0 },
+        { _RK_GLOBAL, "reconnect.backoff.ms", _RK_C_INT,
+          _RK(reconnect_backoff_ms),
+          "The initial time to wait before reconnecting to a broker "
+          "after the connection has been closed. "
+          "The time is increased exponentially until "
+          "`reconnect.backoff.max.ms` is reached. "
+          "+-20% jitter is applied to each reconnect backoff. "
```
Comment (rnpridgeon) [inline src/rdkafka_conf.c]: I personally find this easier to read as one sentence. i.e. 

"The time is increased exponentially until reconnect.backoff.max.ms is reached +-20% jitter."

Comment (edenhill) [reply src/rdkafka_conf.c]: That is more ambigious though, and possibly incorrect:
jitter is only applied to the individual backoff, not to the exponential increase.

Comment (rnpridgeon) [reply src/rdkafka_conf.c]: Fair point

File: src/rdkafka_broker.c [hunk 7]
```diff
@@ -1567,6 +1569,111 @@ int rd_kafka_socket_cb_generic (int domain, int type, int protocol,
...
+
+        /* Apply +-20% jitter to next backoff. */
+        backoff = rd_jitter((int)(float)rkb->rkb_reconnect_backoff_ms * 0.8,
+                            (int)(float)rkb->rkb_reconnect_backoff_ms * 1.2);
+
+         /* Cap to reconnect.backoff.max.ms. */
+        backoff = RD_MIN(backoff, conf->reconnect_backoff_max_ms);
+
+        /* Set time of next reconnect */
+        rkb->rkb_ts_reconnect = now + (backoff * 1000);
```
Comment (cmccabe) [inline src/rdkafka_broker.c]: Can we name this rkb_ts_reconnect_us to make it more clear that this is in microseconds?  Otherwise, this is kind of confusing.

File: src/rdkafka_broker.c [hunk 8]
```diff
@@ -1567,6 +1569,111 @@ int rd_kafka_socket_cb_generic (int domain, int type, int protocol,
...
+        int backoff;
+
+        /* If last connection attempt was more than reconnect.backoff.max.ms
+         * ago, reset the reconnect backoff to the initial
+         * reconnect.backoff.ms value.
+         * Or if the current backoff increment has reached the max value. */
+        if (rkb->rkb_ts_reconnect + (conf->reconnect_backoff_max_ms * 1000) <
+            now ||
+            rkb->rkb_reconnect_backoff_ms >= conf->reconnect_backoff_max_ms)
+                rkb->rkb_reconnect_backoff_ms = conf->reconnect_backoff_ms;
```
Comment (cmccabe) [inline src/rdkafka_broker.c]: Do we really want to reset to the minimum backoff once enough time has elapsed?  It seems better to keep the backoff high until we have some contact with the server.

File: src/rdkafka_broker.c (hunk 7 above)
Comment (edenhill) [reply src/rdkafka_broker.c]: Good point, but the `ts_` prefix convention is used throughout the code to indicate a microsecond rd_ts_t absolute timestamp:

```
$ git grep 'rd_ts_t.*ts_.*;' src/*.h
src/rdinterval.h:        rd_ts_t    ri_ts_last; /* last interval timestamp */
src/rdkafka_broker.h:        rd_ts_t             rkb_ts_rsal_last;
src/rdkafka_broker.h:   rd_ts_t             rkb_ts_fetch_backoff;
src/rdkafka_broker.h:        rd_ts_t             rkb_ts_state;        /* Timestamp of last
src/rdkafka_broker.h:        rd_ts_t             rkb_ts_tx_last;    /**< Timestamp of last
src/rdkafka_broker.h:   rd_ts_t             rkb_ts_metadata_poll; /* Next metadata poll time */
src/rdkafka_broker.h:        rd_ts_t             rkb_ts_reconnect;
src/rdkafka_buf.h:      rd_ts_t rkbuf_ts_retry;    /* Absolute send retry time */
src/rdkafka_buf.h:      rd_ts_t rkbuf_ts_enq;
src/rdkafka_buf.h:      rd_ts_t rkbuf_ts_sent;    /* Initially: Absolute time of transmission,
src/rdkafka_buf.h:        rd_ts_t rkbuf_ts_timeout; /* Request timeout (absolute time). */
src/rdkafka_cgrp.h:        rd_ts_t            rkcg_ts_statechange;     /* Timestamp of last
src/rdkafka_cgrp.h:     rd_ts_t            rkcg_ts_terminate;       /* Timestamp of when
src/rdkafka_cgrp.h:                rd_ts_t            ts_rebalance;       /* Timestamp of
src/rdkafka_int.h:        rd_ts_t          rk_ts_metadata;    /* Timestamp of most recent
src/rdkafka_int.h:      rd_ts_t          rk_ts_full_metadata;       /* Timesstamp of .. */
src/rdkafka_int.h:                rd_ts_t ts_idemp_state;/**< Last state change */
src/rdkafka_metadata.h:        rd_ts_t rkmce_ts_expires;                /* Expire time */
src/rdkafka_metadata.h:        rd_ts_t rkmce_ts_insert;                 /* Insert time */
src/rdkafka_msg.h:                        rd_ts_t ts_timeout; /* Message timeout */
src/rdkafka_msg.h:                        rd_ts_t ts_enq;     /* Enqueue/Produce time */
src/rdkafka_msg.h:                        rd_ts_t ts_backoff; /* Backoff next Produce until
src/rdkafka_op.h:                        rd_ts_t ts_timeout;
src/rdkafka_partition.h:        rd_ts_t            rktp_ts_fetch_backoff; /* Back off fetcher for
src/rdkafka_partition.h:        rd_ts_t            rktp_ts_committed_offset; /* Timestamp of last
src/rdkafka_partition.h:        rd_ts_t            rktp_ts_offset_lag;
src/rdkafka_topic.h:    rd_ts_t            rkt_ts_metadata; /* Timestamp of last metadata
```

File: src/rdkafka_broker.c [hunk 9]
```diff
@@ -1567,6 +1569,111 @@ int rd_kafka_socket_cb_generic (int domain, int type, int protocol,
...
+        rd_ts_t remains;
+
+        if (unlikely(rkb->rkb_ts_reconnect == 0))
+                return 0; /* immediate */
+
+        remains = rkb->rkb_ts_reconnect - now;
+        if (remains <= 0)
+                return 0; /* immediate */
+
+        return (int)(remains / 1000);
```
Comment (cmccabe) [inline src/rdkafka_broker.c]: Can we avoid casting this to int, or at least check for overflow?

Comment (edenhill) [reply src/rdkafka_broker.c]: Relative millisecond timeouts are represented as ints in the code, and there should be no way for this calculation to overflow since the config properties to specify the backoff are limited to 60*60*1000 milliseconds.

File: src/rdkafka_broker.c (hunk 8 above)
Comment (edenhill) [reply src/rdkafka_broker.c]: You're probably right!

File: CONFIGURATION.md [hunk 10]
```diff
@@ -31,7 +31,7 @@ broker.address.ttl                       |  *  | 0 .. 86400000   |          1000
...
 reconnect.backoff.jitter.ms              |  *  | 0 .. 3600000    |             0 | **Deprecated**: No longer used. See `reconnect.backoff.ms` and `reconnect.backoff.max.ms`. <br>*Type: integer*
-reconnect.backoff.ms                     |  *  | 0 .. 3600000    |           100 | The initial time to wait before reconnecting to a broker after the connection has been closed. The time is increased exponentially until `reconnect.backoff.max.ms` is reached. +-20% jitter is applied to each reconnect backoff. A value of 0 disables the backoff and reconnects immediately. <br>*Type: integer*
+reconnect.backoff.ms                     |  *  | 0 .. 3600000    |           100 | The initial time to wait before reconnecting to a broker after the connection has been closed. The time is increased exponentially until `reconnect.backoff.max.ms` is reached. +-20% to +50% jitter is applied to each reconnect backoff. A value of 0 disables the backoff and reconnects immediately. <br>*Type: integer*
```
Comment (mhowlett) [inline CONFIGURATION.md]: `+-20%` -> `-20%`
